### PR TITLE
Introduce DeploymentObserver

### PR DESCRIPTION
**Pull request by xitronix**
* Introduce BlockchainService
* Add ContractWhiteList to PublicRelayerConfig
* Export getContractWhitelist

### CODE REVIEW

File: universal-login-sdk/lib/observers/DeploymentObserver.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import {utils} from 'ethers';
+
+export type OnContractDeployed = (
+  contractAddress: string,
+  bytecode: string
+ ) => void;
+
+export class DeploymentObserver extends ObserverRunner {
+  private onContractDeployed?: OnContractDeployed;
+  private contractAddress?: string;
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/observers/DeploymentObserver.ts]: futureContractAddress

File: universal-login-sdk/lib/observers/DeploymentObserver.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    return () => {
+      this.stop();
+    };
+  }
+
+  async tick() {
+    await this.checkContract(this.contractAddress!);
+  }
+
+  private isContractDeployed(bytecode: string) {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/observers/DeploymentObserver.ts]: Move to commons:
```ts
contractExist(bytecode: string) {
  ensure(bytecode.length > 0, 'Empty bytecode');
  return bytecode !== '0x';
}
```

File: universal-login-sdk/lib/sdk.ts [hunk 3]
```diff
@@ -49,12 +54,18 @@ class UniversalLoginSDK {
     await this.getRelayerConfig();
     const [privateKey, contractAddress] = await createFutureWallet(this.config!.factoryAddress, this.provider);
     this.getBalanceObserver();
+    this.getDeploymentObserver();
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/sdk.ts]: Move to `onBalanceChanged`

File: universal-login-sdk/lib/sdk.ts [hunk 4]
```diff
@@ -49,12 +54,18 @@ class UniversalLoginSDK {
...
     this.balanceObserver!.startAndSubscribe(contractAddress, this.onBalanceChanged);
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/sdk.ts]:
```ts
getFutureWallet(onDeploy, onBalanceChanged?)
...
(address) => this.onBalanceChanged(address, onDeploy, onBalanceChanged)
```

File: universal-login-sdk/lib/sdk.ts [hunk 5]
```diff
@@ -49,12 +54,18 @@ class UniversalLoginSDK {
...
     return [privateKey, contractAddress];
   }
 
-  onBalanceChanged() {
-    /*TODO add relayerApi.balanceChanged() and deploymentObserver.start()*/
+  onBalanceChanged(contractAddress: string) {
+    const onContractDeployed: OnContractDeployed = (
+      contractAdress: string,
+      bytecode: string
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/sdk.ts]: Unnecessary `bytecode`

File: universal-login-sdk/test/observers/DeploymentObserver.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  let deploymentObserver: DeploymentObserver;
+
+  beforeEach(() => {
+    deploymentObserver = new DeploymentObserver(blockchainService as any, contractWhiteList);
+    deploymentObserver.step = 10;
+  });
+
+  afterEach(() => {
+    blockchainService.getCode.resetHistory();
+  });
```
Comment (marekkirejczyk) [inline universal-login-sdk/test/observers/DeploymentObserver.ts]: Push to the bottom of the file

File: universal-login-sdk/lib/observers/DeploymentObserver.ts (hunk 1 above)
Comment (xitronix) [reply universal-login-sdk/lib/observers/DeploymentObserver.ts]: ✔️

File: universal-login-sdk/lib/sdk.ts (hunk 5 above)
Comment (xitronix) [reply universal-login-sdk/lib/sdk.ts]: ✔️

File: universal-login-sdk/test/observers/DeploymentObserver.ts (hunk 6 above)
Comment (xitronix) [reply universal-login-sdk/test/observers/DeploymentObserver.ts]: ✔️

File: universal-login-sdk/lib/sdk.ts (hunk 3 above)
Comment (xitronix) [reply universal-login-sdk/lib/sdk.ts]: ✔️

**Pull request merged**